File: lib/shared/models/deck.dart
import 'card.dart';

class Deck {
  final List<GameCard> cards;
  Deck._(this.cards);

  factory Deck.fourPlayer() {
    final cards = <GameCard>[];
    const fullSuits = [Suit.spades, Suit.hearts, Suit.clubs];
    const fullRanks = Rank.values;

    for (final suit in fullSuits) {
      for (final rank in fullRanks) { cards.add(GameCard(suit: suit, rank: rank)); }
    }

    // Diamonds: all ranks except 7
    for (final rank in fullRanks) {
      if (rank != Rank.seven) { cards.add(GameCard(suit: Suit.diamonds, rank: rank)); }
    }

    cards.add(GameCard.joker());
    return Deck._(cards);
  }

  List<List<GameCard>> deal(int playerCount) {
    final shuffled = List<GameCard>.from(cards)..shuffle();
    final cardsPerPlayer = shuffled.length ~/ playerCount;
    return List.generate(
      playerCount,
      (i) => shuffled.sublist(i * cardsPerPlayer, (i + 1) * cardsPerPlayer),
    );
  }
}

File: workers/src/game/deck.ts
import { SuitName, RankName, GameCard } from './types';
import { makeCard, makeJoker } from './card';

const FULL_SUITS: SuitName[] = ['spades', 'hearts', 'clubs'];
const ALL_RANKS: RankName[] = ['ace', 'king', 'queen', 'jack', 'ten', 'nine', 'eight', 'seven'];
const DIAMOND_RANKS: RankName[] = ['ace', 'king', 'queen', 'jack', 'ten', 'nine', 'eight'];

export function buildFourPlayerDeck(): GameCard[] {
  const cards: GameCard[] = [];

  for (const suit of FULL_SUITS) {
    for (const rank of ALL_RANKS) {
      cards.push(makeCard(suit, rank));
    }
  }

  // Diamonds: all ranks except seven
  for (const rank of DIAMOND_RANKS) {
    cards.push(makeCard('diamonds', rank));
  }

  cards.push(makeJoker());
  return cards;
}

function shuffle<T>(arr: T[]): T[] {
  const result = [...arr];
  for (let i = result.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [result[i], result[j]] = [result[j], result[i]];
  }
  return result;
}

export function dealHands(deck: GameCard[]): [GameCard[], GameCard[], GameCard[], GameCard[]] {
  const shuffled = shuffle(deck);
  const cardsPerPlayer = Math.floor(shuffled.length / 4);
  return [
    shuffled.slice(0, cardsPerPlayer),
    shuffled.slice(cardsPerPlayer, cardsPerPlayer * 2),
    shuffled.slice(cardsPerPlayer * 2, cardsPerPlayer * 3),
    shuffled.slice(cardsPerPlayer * 3, cardsPerPlayer * 4),
  ];
}
